lib/ansible/plugins/inventory/aws_ec2.py

```diff
@@ -0,0 +1,635 @@
+                return os.environ.get(env_var)
+
+    def _set_credentials(self, config_data):
+        '''
+            :param config_data: contents of the inventory config file
+        '''
+
+        self.boto_profile = config_data.get('boto_profile')
+        self.aws_access_key_id = config_data.get('aws_access_key_id')
+        self.aws_secret_access_key = config_data.get('aws_secret_access_key')
+        self.aws_security_token = config_data.get('aws_security_token')
+
+        if not self.boto_profile:
+            self.boto_profile = self._find_cred(possible_env_vars=('AWS_PROFILE',
+                                                                   'AWS_DEFAULT_PROFILE'))
+
+        if not self.aws_access_key_id:
+            self.aws_access_key_id = self._find_cred(possible_env_vars=('AWS_ACCESS_KEY_ID',
+                                                                        'AWS_ACCESS_KEY',
+                                                                        'EC2_ACCESS_KEY'))
+
+        if not self.aws_secret_access_key:
+            self.aws_secret_access_key = self._find_cred(possible_env_vars=('AWS_SECRET_ACCESS_KEY',
+                                                                            'AWS_SECRET_KEY',
+                                                                            'EC2_SECRET_KEY'))
+
+        if not self.aws_security_token:
+            self.aws_security_token = self._find_cred(possible_env_vars=('AWS_SECURITY_TOKEN',
+                                                                         'AWS_SESSION_TOKEN',
+                                                                         'EC2_SECURITY_TOKEN'))
+        if not self.boto_profile and not (self.aws_access_key_id and self.aws_secret_access_key):
+            raise AnsibleError("Insufficient boto credentials found. Please provide them in your "
+                               "inventory configuration file or set them as environment variables.")
+
+    def _validate_config(self, loader, path):
+        '''
+            :param loader: an ansible.parsing.dataloader.DataLoader object
+            :param path: the path to the inventory config file
+            :return the contents of the config file
+        '''
+        self.verify_file(path)
+
+        # file is config file
+        try:
+            config_data = self.loader.load_from_file(path)
+        except Exception as e:
+            raise AnsibleParserError(to_native(e))
+
+        if not config_data or config_data.get('plugin') != self.NAME:
+            # this is not my config file
+            raise AnsibleParserError("Not a ec2 inventory plugin configuration file")
+
+        return config_data
+
+    def _set_cache(self, inventory, path):
+        '''
+            :param inventory: an ansible.inventory.data.InventoryData object
+            :param path: the path to the inventory config file
+        '''
+        cache_key = self.get_cache_prefix(path)
```

Comment:
Should cacheability be configurable for each "instance" of the plugin in addition to globally (specifically disable-able if global caching is enabled)?

Reply:
the cache will be by 'plugin/source' initially it will be globally enabled, but i hope to implement per plugin and even per source settings